hoomd/data/typeparam.py

```diff
@@ -110,33 +69,93 @@ def __getattr__(self, attr):
             raise AttributeError("'{}' object has no attribute "
                                  "'{}'".format(type(self), attr))
 
-    def __getitem__(self, key):
-        """Access parameters by key."""
-        return self.param_dict[key]
-
     def __setitem__(self, key, value):
-        """Set parameters by key."""
+        """Set parameters for a given type (or type pair).
+
+        .. rubric:: Examples:
+
+        Index types by name:
+
+        .. skip: next if(not hoomd.version.md_built)
+        .. code-block:: python
+
+            langevin.gamma['A'] = 2.0
+
+        Set parameters for multiple types:
+
+        .. skip: next if(not hoomd.version.md_built)
+        .. code-block:: python
+
+            langevin.gamma[['B', 'C']] = 3.0
+
+        Set type pair parameters with a tuple of names:
+
+        .. skip: next if(not hoomd.version.md_built)
+        .. code-block:: python
+
+            lj.params[('A', 'A')] = dict(epsilon=1.5, sigma=2.0)
+
+        Set parameters for multiple pairs (e.g. ('A', 'B') and ('A', 'C')):
+
+        .. skip: next if(not hoomd.version.md_built)
+        .. code-block:: python
+
+            lj.params[('A', ['B', 'C'])] = dict(epsilon=0, sigma=0)
+
+        Set parameters for multiple pairs (e.g. ('B', 'B'), ('B', 'C'), ('C',
+        'B'), and ('C', 'C')):
+
+        .. skip: next if(not hoomd.version.md_built)
+        .. code-block:: python
+
+            lj.params[(['B', 'C'], ['B', 'C'])] = dict(epsilon=1, sigma=1)
+
+        Note:
+            Setting the value for *(a,b)* automatically sets the symmetric
+            *(b,a)* parameter to the same value.
+        """
         self.param_dict[key] = value
 
+    def __getitem__(self, key):
```

Comment:
Maybe we should mention that `__getitem__` supports the same indexing logic shown above?

Reply:
Isn't this a weird behavior? Do other mapping types in the Python ecosystem follow this behavior?
```
>>> lj.params[('A', 'A')] = dict(epsilon=1, sigma=2)
>>> lj.params[('A', 'B')] = dict(epsilon=2, sigma=3)
>>> lj.params[('B', 'B')] = dict(epsilon=3, sigma=4)
>>> lj.params[('A', 'A')]
_HOOMDDict{'epsilon': 1.0, 'sigma': 2.0}
>>> lj.params[('A', ['A', 'B'])]
{('A', 'A'): _HOOMDDict{'epsilon': 1.0, 'sigma': 2.0}, ('A', 'B'): _HOOMDDict{'epsilon': 2.0, 'sigma': 3.0}}
```
I see that with a single pair, it returns a single dict. With smart indexing, it returns a dict of dicts. What use-case does this enable for users and how would I demonstrate it in an example? 

This seems very awkard:
```
>>> lj.params[('A', ['A', 'B'])][('A', 'A')]
_HOOMDDict{'epsilon': 1.0, 'sigma': 2.0}
```

Reply:
Most mapping types I am aware of do not support setting or getting in this abbreviated fashion. We only support it because of the setting which I thought was useful and somewhat intuitive when shown. I figured it would be confusing to only support the smart indexing for setting, so instead it is supported on any operation that takes in keys.

Reply:
I see. I find it confusing that setting with multiple keys broadcasts a single dictionary to those values where getting with multiple keys returns a dictionary with multiple results. I cannot think of any use-case to demonstrate for users, so I added documentation that states multiple keys are allowed in `__getitem__` and show an equivalent notation to explain the output:
```
gammas = langevin.gamma[['A', 'B']]
gammas = {key: langevin.gamma[key] for key in ['A', 'B']}
```